### PR TITLE
[release/7.0] Improve error message for ByRef parameter exception

### PR DESCRIPTION
**Pull request by brunolins16**
Fixes https://github.com/dotnet/aspnetcore/issues/43256

## Description

The mechanism to detect Try/Parse method fails when a `ByRef` type with a very obscure message.

```
System.TypeLoadException: Could not load type 'System.Char&' from assembly 'System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e'.
   at System.RuntimeTypeHandle.MakeByRef()
   at System.RuntimeType.MakeByRefType()
   at Microsoft.AspNetCore.Http.ParameterBindingMethodCache.<FindTryParseMethod>g__Finder|15_0(Type type)
```

This change detects if it is a by ref type and:

- Minimal APIs: throws a better error
- MVC: avoid the obscure exception and keep the .NET 6 behavior.

## Customer Impact

Minimal APIs is a very small impact, since it is just an improved error message. In MVC we are avoiding the obscure error and reporting as same as we have in .NET 6.

## Regression?

- [ ] Yes
- [x] No

Not really, the error message is very obscure and does not explain what is wrong.

## Risk

- [ ] High
- [ ] Medium
- [x] Low

Very low risk, we're just fixing the error message.

## Verification

- [X] Manual (required)
- [x] Automated

## Packaging changes reviewed?

- [ ] Yes
- [ ] No
- [x] N/A

### CODE REVIEW

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 1]
```diff
@@ -645,6 +645,22 @@ private static Expression CreateArgument(ParameterInfo parameter, RequestDelegat
...
+            var attribute = string.Empty;
+
+            if (parameter.Attributes.HasFlag(ParameterAttributes.In))
+            {
+                attribute = "in ";
+            }
+            else if (parameter.Attributes.HasFlag(ParameterAttributes.Out))
+            {
+                attribute = "out ";
+            }
```
Comment (ReubenBond) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:659]: Should there be a "ref" alternative here?

Comment (halter73) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs:659]: Probably. Should that just be the default instead of ""?

Comment (brunolins16) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs:659]: I added "by reference" in the error message but why not.  Let me update it.

**Review by halter73 — APPROVED** [review]

**Comment by Pilchie** [comment]
Approved for .NET 7 RC2

**Comment by brunolins16** [comment]
@dotnet/aspnet-build Merge please.

**Pull request merged**